File: src/shared/utils/generate-pagination-metadata.ts
import { IPaginationMetadata } from '../domain/interfaces/pagination-metadata.interface';
import { QueryParamsValue } from '../types/query-params-value';

function generatePaginationMetaLinks(
  baseUrl: string,
  currentPage: number,
  totalItems: number,
  pageSize: number,
  queryParams?: Record<string, QueryParamsValue>,
): {
  first: string | null;
  prev: string | null;
  next: string | null;
  last: string | null;
} {
  const totalPages = Math.ceil(totalItems / pageSize);

  const getQueryString = (params: Record<string, QueryParamsValue>): string => {
    return Object.entries(params)
      .filter(([, value]) => value !== undefined)
      .map(([key, value]) => {
        if (Array.isArray(value)) {
          return value.map((item) => `${key}=${item}`).join('&');
        } else {
          return `${key}=${value}`;
        }
      })
      .join('&');
  };

  const getLink = (page: number): string | null => {
    if (page < 1 || page > totalPages) {
      return null;
    }

    const params: Record<string, string> = {
      page: page.toString(),
      pageSize: pageSize.toString(),
      ...queryParams,
    };

    const queryString = getQueryString(params);
    return `${baseUrl}?${queryString}`;
  };

  const first = getLink(1);
  const prev = getLink(currentPage - 1);
  const next = getLink(currentPage + 1);
  const last = getLink(totalPages);

  return { first, prev, next, last };
}

export function generatePaginationMeta(
  baseUrl: string,
  totalItems: number,
  currentPage: number,
  pageSize: number,
  queryParams?: Record<string, QueryParamsValue>,
): IPaginationMetadata {
  const totalPages = Math.ceil(totalItems / pageSize);

  const links = generatePaginationMetaLinks(
    baseUrl,
    currentPage,
    totalItems,
    pageSize,
    queryParams,
  );

  return {
    page: currentPage,
    pageSize: pageSize,
    totalItems: totalItems,
    totalPages: totalPages,
    links: links,
  };
}

File: src/shared/utils/generate_pagination_metadata.py
from math import ceil
from typing import Dict, Optional
from src.shared.types.query_params_value import QueryParamsValue
from src.shared.domain.interfaces.pagination_metadata_interface import (
    IPaginationMetadata,
)


def generate_pagination_meta_links(
    base_url: str,
    current_page: int,
    total_items: int,
    page_size: int,
    query_params: Optional[Dict[str, QueryParamsValue]] = None,
) -> Dict[str, Optional[str]]:
    total_pages = ceil(total_items / page_size)

    def get_query_string(params: Dict[str, QueryParamsValue]) -> str:
        query_parts = []
        for key, value in params.items():
            if value is not None:
                if isinstance(value, list):
                    query_parts.extend(f"{key}={item}" for item in value)
                else:
                    query_parts.append(f"{key}={value}")
        return "&".join(query_parts)

    def get_link(page: int) -> Optional[str]:
        if page < 1 or page > total_pages:
            return None
        params = {"page": str(page), "page_size": str(page_size)}
        if query_params:
            params.update({k: v for k, v in query_params.items() if v is not None})
        query_string = get_query_string(params)
        return f"{base_url}?{query_string}"

    return {
        "first": get_link(1),
        "prev": get_link(current_page - 1),
        "next": get_link(current_page + 1),
        "last": get_link(total_pages),
    }


def generate_pagination_meta(
    base_url: str,
    total_items: int,
    current_page: int,
    page_size: int,
    query_params: Optional[Dict[str, QueryParamsValue]] = None,
) -> IPaginationMetadata:
    total_pages = ceil(total_items / page_size)

    links = generate_pagination_meta_links(
        base_url, current_page, total_items, page_size, query_params
    )

    return {
        "page": current_page,
        "page_size": page_size,
        "total_items": total_items,
        "total_pages": total_pages,
        "links": links,
    }
